File: 513.找树左下角的值.c
/*
 * @lc app=leetcode.cn id=513 lang=c
 *
 * [513] 找树左下角的值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int res;
int maxHeight;
void back(struct TreeNode *root, int depth)
{
  if (root == NULL)
    return;
  if (root->left == NULL && root->right == NULL)
  {
    if (depth > maxHeight) // 通过最大深度第一次增加判断该叶子节点是在最左边
    {
      maxHeight = depth;
      res = root->val;
    }
    return;
  }
  back(root->left, depth + 1);
  back(root->right, depth + 1);
}

int findBottomLeftValue(struct TreeNode *root)
{
  // 法一：递归
  maxHeight = 1;
  res = root->val;
  if (root != NULL)
    back(root, 1);
  return res;
}
// @lc code=end

File: store/513.找树左下角的值.js
/*
 * @lc app=leetcode.cn id=513 lang=javascript
 *
 * [513] 找树左下角的值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let maxDepth;
let val;
const getLeft = (node, depth) => {
  if (!node) return;
  getLeft(node.left, depth + 1);
  if (maxDepth < depth) {
    val = node.val;
    maxDepth = depth;
  }
  getLeft(node.right, depth + 1);
};
var findBottomLeftValue = function (root) {
  maxDepth = 0;
  val = -1;
  getLeft(root, 1);
  return val;
};
// @lc code=end
